### PR TITLE
Add ability to intercept streams with middleware

### PR DESCRIPTION
**Pull request by mgazza**
Add the ability to add support for such things as zmodem and/or other
    stream middleware's

### CODE REVIEW

File: term.go [hunk 1]
```diff
@@ -497,3 +504,20 @@ func (t *Terminal) Dragged(d *fyne.DragEvent) {
 func (t *Terminal) DragEnd() {
 	t.selecting = false
 }
+
+// SetMiddleware sets the middleware function that will be used when creating a new terminal.
+// The middleware function is responsible for setting up the I/O readers and writers.
```
Comment (andydotxyz) [inline term.go:509]: The way middleware is implemented it is an interface type not function. 
It feels like a function would be enough to set up the interception?

**Review by andydotxyz — COMMENTED** [review]
I know this is a common concept in HTTP handling, though the implementation doesn't seem to quite match by introducing the requirement to implement types instead of just a function?
Also is the "SetMiddleware" a clear enough name? In the realm of terminals or GUIs I'm not sure it`s totally clear...

File: term.go (hunk 1 above)
Comment (mgazza) [reply term.go:509]: thats the beauty of golang right?
```
// MiddlewareFunc is a function type that matches the signature of the
// SetupMiddleware method in the Middleware interface.
type MiddlewareFunc func(r io.Reader, w io.WriteCloser) (io.Reader, io.WriteCloser)

// SetupMiddleware allows MiddlewareFunc to satisfy the Middleware interface.
// It calls the MiddlewareFunc itself.
func (m MiddlewareFunc) SetupMiddleware(r io.Reader, w io.WriteCloser) (io.Reader, io.WriteCloser) {
	return m(r, w)
}
```

**Comment by mgazza** [comment]
> Sorry for the test failures, fixed on master

no worries rebased

File: term.go (hunk 1 above)
Comment (andydotxyz) [reply term.go:509]: I guess if following the http middleware / handler idea it was missing the func helper? https://pkg.go.dev/net/http#HandlerFunc

**Comment by andydotxyz** [comment]
> Also is the "SetMiddleware" a clear enough name? In the realm of terminals or GUIs I'm not sure it`s totally clear...

I think this is still niggling me - it's such a generic term that I have never heard use in stream handling or terminals before.

File: term.go (hunk 1 above)
Comment (mgazza) [reply term.go]: yes essentially the same pattern

**Review by andydotxyz — APPROVED** [review]
Thanks, I think that is clear now

**Pull request merged**